sklearn/decomposition/dict_learning.py

```diff
@@ -343,35 +352,32 @@ def _update_dict(dictionary, Y, code, verbose=False, return_r2=False,
         Updated dictionary.
 
     """
-    n_components = len(code)
-    n_samples = Y.shape[0]
+    n_features, n_components = B.shape
     random_state = check_random_state(random_state)
     # Residuals, computed 'in-place' for efficiency
-    R = -np.dot(dictionary, code)
-    R += Y
+    R = -np.dot(dictionary, A)
+    R += B
     R = np.asfortranarray(R)
-    ger, = linalg.get_blas_funcs(('ger',), (dictionary, code))
+    ger, = linalg.get_blas_funcs(('ger',), (dictionary, A))
     for k in range(n_components):
         # R <- 1.0 * U_k * V_k^T + R
-        R = ger(1.0, dictionary[:, k], code[k, :], a=R, overwrite_a=True)
-        dictionary[:, k] = np.dot(R, code[k, :].T)
+        R = ger(1.0, dictionary[:, k], A[k, :], a=R, overwrite_a=True)
+        dictionary[:, k] = np.dot(R, A[k, :])
```

Comment:
I just checked with with your code (modified)
```
import time
import numpy as np
from sklearn.decomposition.dict_learning import dict_learning

rng = np.random.RandomState(0)
X = rng.randn(30, 20)

t0 = time.time()
U, V, _ = dict_learning(X, n_components=3, alpha=0.1, random_state=0,
                        verbose=2, max_iter=10000, tol=1e-20)
print(0.5 * np.sum((X - np.dot(U, V)) ** 2) + 0.1 * np.sum(np.abs(U)))
print("Time elapsed: %.2f" % (time.time() - t0))
```
 ** On master:**

```
Iteration  647 (elapsed time:  16s,  0.3mn, current cost  198.777)
Iteration  648 (elapsed time:  16s,  0.3mn, current cost  198.777)
--- Convergence reached after 648 iterations
198.776663706
Time elapsed: 16.46
```

**On PR:**
```
--- Convergence reached after 647 iterations
Learning code... done (total time:  10s,  0.2mn)
198.776663706
Time elapsed: 10.38
```

Thus same, results but my PR is  faster than the master branch. Can you confirm. Thanks.